x/valset-pref/validator_set_test.go

```diff
@@ -232,6 +232,111 @@ func (s *KeeperTestSuite) TestUndelegateFromValidatorSet() {
+			delegateAmt:           []osmomath.Int{sdk.NewInt(10), sdk.NewInt(90)},
+			undelegateAmt:         sdk.NewInt(50),
+			expectedUndelegateAmt: []osmomath.Int{sdk.NewInt(5), sdk.NewInt(45)},
+		},
+		{
+			name:                  "undelegate all, weights do not match the current delegations to valset",
+			delegateAmt:           []osmomath.Int{sdk.NewInt(90), sdk.NewInt(10)},
+			undelegateAmt:         sdk.NewInt(100),
+			expectedUndelegateAmt: []osmomath.Int{sdk.NewInt(90), sdk.NewInt(10)},
+		},
+		{
+			name:                  "undelegate some, weights do not match the current delegations to valset",
+			delegateAmt:           []osmomath.Int{sdk.NewInt(90), sdk.NewInt(10)},
+			undelegateAmt:         sdk.NewInt(50),
+			expectedUndelegateAmt: []osmomath.Int{sdk.NewInt(45), sdk.NewInt(5)},
+		},
+		{
+			name:          "error: attempt to undelegate more than delegated",
+			delegateAmt:   []osmomath.Int{sdk.NewInt(100), sdk.NewInt(50)},
+			undelegateAmt: sdk.NewInt(200),
+			expectedError: types.UndelegateMoreThanDelegatedError{TotalDelegatedAmt: sdk.NewDec(150), UndelegationAmt: sdk.NewInt(200)},
+		},
+		{
+			name:          "error: user does not have val-set preference set",
+			delegateAmt:   []osmomath.Int{sdk.NewInt(100), sdk.NewInt(50)},
+			undelegateAmt: sdk.NewInt(100),
+			noValset:      true,
+			expectedError: types.NoValidatorSetOrExistingDelegationsError{DelegatorAddr: s.TestAccs[0].String()},
+		},
+	}
+	for _, test := range tests {
+		s.Run(test.name, func() {
+			s.SetupTest()
+			valAddrs := s.SetupMultipleValidators(3)
+			defaultDelegator := s.TestAccs[0]
+			bondDenom := s.App.StakingKeeper.BondDenom(s.Ctx)
+
+			// set val-set pref
+			valPreferences := []types.ValidatorPreference{
+				{
+					ValOperAddress: valAddrs[0],
+					Weight:         sdk.NewDecWithPrec(1, 1),
+				},
+				{
+					ValOperAddress: valAddrs[1],
+					Weight:         sdk.NewDecWithPrec(9, 1),
+				},
+			}
+
+			if !test.noValset {
+				s.App.ValidatorSetPreferenceKeeper.SetValidatorSetPreferences(s.Ctx, defaultDelegator.String(), types.ValidatorSetPreferences{
+					Preferences: valPreferences,
+				})
+				// delegate for each of the validators
+				for i, valsetPref := range valPreferences {
+					valAddr, err := sdk.ValAddressFromBech32(valsetPref.ValOperAddress)
+					s.Require().NoError(err)
+					validator, found := s.App.StakingKeeper.GetValidator(s.Ctx, valAddr)
+					s.Require().True(found)
+
+					s.FundAcc(defaultDelegator, sdk.NewCoins(sdk.NewCoin(bondDenom, test.delegateAmt[i])))
+					_, err = s.App.StakingKeeper.Delegate(s.Ctx, defaultDelegator, test.delegateAmt[i], stakingtypes.Unbonded, validator, true)
+					s.Require().NoError(err)
+				}
+			}
+
+			// System Under Test
+			err := s.App.ValidatorSetPreferenceKeeper.UndelegateFromRebalancedValidatorSet(s.Ctx, defaultDelegator.String(), sdk.NewCoin(bondDenom, test.undelegateAmt))
```

Comment:
Don't want to block this PR, but I think we're actually not testing the main important thing here which is that directly delegating/undelegating from a validator changes the weights when the new message is called.

E.g. if it starts at 0.1/0.9 with 10 tokens in validator A and 90 in validator B, then when you delegate 100 tokens to validator A, then withdrawing rebalanced should yield a distribution of 0.55/0.45 (since now validator A has 110 tokens out of 200 and validator B has 90 out of 200)

Would it be possible for us to add tests covering these? I think this could potentially be done with an additional test field and putting system under test into a loop so it runs multiple times with different conditions.